src/org/zaproxy/zap/extension/websocket/manualsend/WebSocketReopenSend.java

```diff
@@ -0,0 +1,388 @@
+    private HashMap<String, Stack<Integer>> reopenedPool;
+    public final static String SEC_WEBSOCKET_KEY = "Sec-WebSocket-Key";
+
+    private ExtensionHistory extensionHistory;
+
+    public WebSocketReopenSend(){
+        persistentConnectionListeners = new ArrayList<>();
+        webSocketReopenListeners = new ArrayList<>();
+        reopenedPool= new HashMap<>();
+    }
+
+    public void addPersistentConnectionListener(PersistentConnectionListener listener) {
+        persistentConnectionListeners.add(listener);
+    }
+
+    public void removePersistentConnectionListener(PersistentConnectionListener listener) {
+        persistentConnectionListeners.remove(listener);
+    }
+
+    public void addWebSocketReopenListeners(WebSocketReopenListener webSocketReopenListener){
+        webSocketReopenListeners.add(webSocketReopenListener);
+    }
+
+    public void removeWebSocketReopenListeners(WebSocketReopenListener webSocketReopenListener){
+        webSocketReopenListeners.remove(webSocketReopenListener);
+    }
+
+    /**
+     * Open a new connection same as closedWebSocketProxy by sending same handshake.
+     * External listener is set at {@link WebSocketReopenSend#addPersistentConnectionListener(PersistentConnectionListener)}
+     * is able to hear and establish new WebSocket connection by HTTP response.
+     * @param webSocketProxy WebSocket Connection is going to re-open (Establish new connection same with that)
+     * @param newSecWebSocketKey New Sec-WebSocket-Key, null to use the same as the initial request
+     * @param followRedirects true to follow redirections
+     * @param useTrackingSessionState false for stateless http request
+     * @return false if doesn't find the http handshake.
+     */
+    public boolean reEstablishConnection(WebSocketProxy webSocketProxy, String newSecWebSocketKey, boolean useTrackingSessionState, boolean followRedirects) {
+
+        this.useTrackingSessionState = useTrackingSessionState;
+        this.followRedirects = followRedirects;
+        String secWebSocketKey;
+
+        if(webSocketProxy instanceof WebSocketProxyV13){
+            try {
+                HttpMessage newHttpHandshake;
+                if( newSecWebSocketKey != null){
+                    secWebSocketKey = newSecWebSocketKey;
+                    newHttpHandshake = webSocketProxy.getHandshakeReference().getHttpMessage().cloneRequest();
+                    newHttpHandshake.getRequestHeader().setHeader(SEC_WEBSOCKET_KEY,secWebSocketKey);
+                }else{
+                    newHttpHandshake = webSocketProxy.getHandshakeReference().getHttpMessage().cloneRequest();
+                    secWebSocketKey = newHttpHandshake.getRequestHeader().getHeader(SEC_WEBSOCKET_KEY);
+                }
+                addToReopenedPool(secWebSocketKey);
+                send(newHttpHandshake);
+            } catch (HttpMalformedHeaderException e) {
+                LOGGER.error(e.getStackTrace().toString());
```

Comment:
To have the exception's stack trace properly logged it should be used other method (e.g. `LOGGER.error(e, e);`).